akka-persistence-typed/src/main/scala/akka/persistence/typed/ExpectingReply.scala

```diff
@@ -11,6 +11,6 @@ import akka.actor.typed.ActorRef
  *
  * @tparam ReplyMessage The type of the reply message
  */
-trait ExpectingReply[ReplyMessage] {
+trait ExpectingReply[-ReplyMessage] {
```

Comment:
I have tried a few things, without much success. I'm not sure what specific usage we are trying to improve with this. The `Behavior[AccountCommand[_]]` pattern seems sufficient to me. It's the reply type of the concrete commands that are interesting, not that all `AccountCommand`s have a reply type of `AccountCommandReply`.

Anyway, I tried an attempt to express that all `AccountCommand`s have a reply type of `AccountCommandReply`. Something like a967830c18 but that doesn't work out.

Reply:
Note, if you do further experiments I'd recommend to do that on top of https://github.com/akka/akka/pull/26813

Reply:
> I have tried a few things

What is it that you're trying to do in your experiments?  Put another way, what is it you don't like about what I'm proposing?

`ActorRef` is contravariant in the type `T` message type it can consume.  `ExpectingReply`'s `ReplyMessage` type has no covariant usages so it isn't forced into being invariant.  My proposal here is to propagates the contravariant nature of `ReplyMessage`.

"Seems sufficient", to me, seems like a weak reason to oppose.  There is test code attached to this change that shows how you can more easily and more safely reuse code and values.

Let me know when #26813 is stable enough to rebase on and I'd be happy to do so.

Reply:
> What is it that you're trying to do in your experiments? Put another way, what is it you don't like about what I'm proposing?

I tried to use your approach on top of the latest samples in #26813. The difference is that they are using the `EventSourcedEntity.withEnforcedReplies` and the `EntityTypeKey`. That is the primary usage with with Cluster Sharding, which is what we should focus on.

> ActorRef is contravariant in the type T message type it can consume. ExpectingReply's ReplyMessage type has no covariant usages so it isn't forced into being invariant. My proposal here is to propagates the contravariant nature of ReplyMessage.

It's probably completely correct to make it contravariant. I'm not opposed to that. I'm more interested in what we would use it for and how to show that in documentation.

>  There is test code attached to this change that shows how you can more easily and more safely reuse code and values.

Commented there

> Let me know when #26813 is stable enough to rebase on and I'd be happy to do so.

It's stable enough. It's important to try these things with `EventSourcedEntity.withEnforcedReplies`.